### ISSUE TITLE
IndexerProperty can't have a null value in one of its columns

### ISSUE THREAD

**Issue by YeskaNova** [dotnet/efcore#20718]
If only one column of an indexerProperty is null, then the whole Dictionary will be null.

In the given example, both assertions should be OK, because both users have at least the "a" key set:

``` C#
                Debug.Assert(users[0].Data != null);
                Debug.Assert(users[1].Data != null);
```

 There is a bug that makes the second user.Data equals null, only because "b" key has no value.

What steps can we follow to reproduce the issue?

``` C#
using Microsoft.EntityFrameworkCore;
using Microsoft.Extensions.DependencyInjection;
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;

public class User
{
    public short Id { get; set; }
    public Dictionary<string, object> Data { get; set; }
}
public class SqliteDbContext : DbContext
{
    public ISet<User> Users { get; set; }

    public SqliteDbContext(DbContextOptions options) : base(options)
    {
    }

    protected override void OnModelCreating(ModelBuilder modelBuilder)
    {
        var etb = modelBuilder.Entity<User>();
        etb.HasKey(x => x.Id);
        etb.OwnsOne(x => x.Data, x =>
          {
              x.IndexerProperty(typeof(int), "a");
              x.IndexerProperty(typeof(string), "b");
          });

    }
}
class Program
{
    static void Main(string[] args)
    {
        var sc = new ServiceCollection();
        var guid = Guid.NewGuid();
        sc.AddDbContext<SqliteDbContext>(op => op.UseSqlite("Data Source=" + guid + ".db;"));
        using (var provider = sc.BuildServiceProvider())
        {
            using (var scope = provider.CreateScope())
            using (var dbcontext = scope.ServiceProvider.GetService<SqliteDbContext>())
            {
                dbcontext.Database.EnsureCreated();
                dbcontext.Set<User>().Add(new User() { Data = new Dictionary<string, object>() { { "a", 1 }, { "b", "1" } } });
                dbcontext.Set<User>().Add(new User() { Data = new Dictionary<string, object>() { { "a", 2 }, { "b", null } } });
                dbcontext.SaveChanges();
            }
            using (var scope = provider.CreateScope())
            using (var dbcontext = scope.ServiceProvider.GetService<SqliteDbContext>())
            {
                var users = dbcontext.Set<User>().Where(x => x.Id > 0).ToArray();
                Debug.Assert(users[0].Data != null);
                Debug.Assert(users[1].Data != null);
            }
        }
    }
}
```

### Further technical details

EF Core version: 5.0.0-preview.5.20222.1
Database provider: (e.g. Microsoft.EntityFrameworkCore.SqlServer) issue in both
Target framework: (e.g. .NET Core 3.0) 5.0
Operating system: Windows 10 Pro
IDE: (e.g. Visual Studio 2019 16.3) VS 2019

**Comment by smitpatel** [comment]
NonNullableReferencePropertyConvention is marking the property b as required even though it is of type string. Hence query does not materialize the instance since all required properties does not have non-null values.

**Comment by roji** [comment]
@smitpatel do you mean this is reproducible even when nullability isn't turned on in the project/source?

**Comment by smitpatel** [comment]
Yes.

**Comment by roji** [comment]
This is because the Dictionary type is a annotated for nullability (as part of the BCL), and our NRT logic doesn't currently take into account the extra stuff needed for generic types (i.e. interpreting the constraints). FWIW ASP.NET also skips this case for now (https://github.com/dotnet/aspnetcore/blob/master/src/Mvc/Mvc.DataAnnotations/src/DataAnnotationsMetadataProvider.cs#L492) since it's not trivial.

#29723 is still supposed to provide a standard reflection API to provide all the relevant nullability information, so I don't think we should rush to implement this ourselves (not yet, anyway). We should probably implement the ASP.NET "solution" of ignoring context on generic types to unblock people, and reevaluate later in the release.

**Comment by roji** [comment]
Note that the current "fix" (#20823) will completely ignore context information on generic types. This blocks some legitimate usage, i.e. where a user generic type has a non-generic property that needs to inherit the NRT context. We would ideally fix this by 5.0 by using a new API from https://github.com/dotnet/runtime/issues/29723, or possibly by implementing the proper differentiation ourselves (ignoring only *generic* properties on the generic type).
